File: server/clients/mpc_client.py
"""
MPC-BE web interface client.

Setup on HTPC
-------------
MPC-BE → View → Options → Player → Web Interface
  ☑ Listen on port:  13579
  (leave "Allow access from localhost only" UNCHECKED so the iPad can reach it)

API endpoints used
------------------
GET /variables.html   → current player state (JSON or legacy JS-variable format)
GET /command.html     → send a command via ?wm_command=N (and optional params)

Key command IDs
---------------
  887  Play/Pause toggle
  888  Stop
  891  Play
  892  Pause
  907  Volume +5%
  908  Volume -5%
  909  Mute toggle
  889  Seek  (also pass &position=<milliseconds>)
   -1  Open file  (also pass &path=<url-encoded Windows path>)
"""
from __future__ import annotations

import logging
import re
from typing import Any, Optional
import urllib.parse

import httpx

logger = logging.getLogger(__name__)


class MPCStatus:
    def __init__(self, data: dict, reachable: bool = True):
        self._d       = data
        self.reachable = reachable

    @property
    def file(self) -> str:
        return self._d.get("file", "") or self._d.get("filepath", "")

    @property
    def filename(self) -> str:
        return self._d.get("filename", "") or (self.file.split("\\")[-1] if "\\" in self.file else self.file.split("/")[-1])

    @property
    def state(self) -> int:
        """0=stopped, 1=paused, 2=playing"""
        v = self._d.get("state", 0)
        try:
            return int(v)
        except (ValueError, TypeError):
            return 0

    @property
    def is_playing(self) -> bool:
        return self.state == 2

    @property
    def is_paused(self) -> bool:
        return self.state == 1

    @property
    def position_ms(self) -> int:
        v = self._d.get("position", 0)
        try:
            return int(v)
        except (ValueError, TypeError):
            return 0

    @property
    def duration_ms(self) -> int:
        v = self._d.get("duration", 0)
        try:
            return int(v)
        except (ValueError, TypeError):
            return 0

    @property
    def position_str(self) -> str:
        return self._d.get("positionstring", _ms_to_str(self.position_ms))

    @property
    def duration_str(self) -> str:
        return self._d.get("durationstring", _ms_to_str(self.duration_ms))

    @property
    def volume(self) -> int:
        v = self._d.get("volumelevel", 100)
        try:
            return int(v)
        except (ValueError, TypeError):
            return 100

    @property
    def muted(self) -> bool:
        v = self._d.get("muted", False)
        if isinstance(v, bool):
            return v
        return str(v).lower() in ("1", "true", "yes")

    def to_dict(self) -> dict:
        return {
            "reachable":   self.reachable,
            "file":        self.file,
            "filename":    self.filename,
            "state":       self.state,
            "is_playing":  self.is_playing,
            "is_paused":   self.is_paused,
            "position_ms": self.position_ms,
            "duration_ms": self.duration_ms,
            "position_str": self.position_str,
            "duration_str": self.duration_str,
            "volume":      self.volume,
            "muted":       self.muted,
        }


class MPCClient:
    def __init__(self, base_url: str):
        self._base = base_url.rstrip("/")

    async def get_status(self) -> MPCStatus:
        try:
            async with httpx.AsyncClient(timeout=3) as c:
                r = await c.get(f"{self._base}/variables.html")
                r.raise_for_status()
                parsed = self._parse_variables(r.text)
                return MPCStatus(parsed, reachable=True)
        except Exception as exc:
            logger.debug("MPC-BE unreachable at %s: %s", self._base, exc)
            return MPCStatus({}, reachable=False)

    async def command(self, wm_command: int, **extra_params) -> bool:
        params: dict[str, Any] = {"wm_command": wm_command, **extra_params}
        try:
            async with httpx.AsyncClient(timeout=3) as c:
                r = await c.get(f"{self._base}/command.html", params=params)
                return r.status_code in (200, 201, 204)
        except Exception as exc:
            logger.debug("MPC-BE command %d failed: %s", wm_command, exc)
            return False

    async def play_pause(self) -> bool:
        return await self.command(887)

    async def play(self) -> bool:
        return await self.command(891)

    async def pause(self) -> bool:
        return await self.command(892)

    async def stop(self) -> bool:
        return await self.command(888)

    async def mute(self) -> bool:
        return await self.command(909)

    async def volume_up(self) -> bool:
        return await self.command(907)

    async def volume_down(self) -> bool:
        return await self.command(908)

    async def seek(self, position_ms: int) -> bool:
        return await self.command(889, position=position_ms)

    async def ping(self) -> bool:
        """Return True if MPC-BE's web interface is reachable (i.e. the process is running)."""
        try:
            async with httpx.AsyncClient(timeout=2) as c:
                r = await c.get(f"{self._base}/variables.html")
                return r.status_code == 200
        except Exception:
            return False

    async def open_file(self, path: str) -> bool:
        """
        Open a file in MPC-BE.
        *path* should be the local Windows path on the HTPC, e.g.
        'D:\\Media\\Movies\\Inception (2010)\\Inception (2010).mkv'
        """
        try:
            async with httpx.AsyncClient(timeout=5) as c:
                r = await c.get(
                    f"{self._base}/command.html",
                    params={"wm_command": -1, "path": path},
                )
                return r.status_code in (200, 201, 204)
        except Exception as exc:
            logger.warning("MPC-BE open_file failed: %s", exc)
            return False

    # ------------------------------------------------------------------
    # Variable parsing (handles both JSON and legacy JS format)
    # ------------------------------------------------------------------

    @staticmethod
    def _parse_variables(text: str) -> dict:
        # Try JSON first (newer MPC-BE versions)
        stripped = text.strip()
        if stripped.startswith("{"):
            try:
                import json
                return json.loads(stripped)
            except Exception:
                pass

        # Legacy JS format: OnVariable("key","value");
        result: dict = {}
        for m in re.finditer(r'OnVariable\("([^"]+)","([^"]*)"\)', text):
            result[m.group(1)] = m.group(2)
        if result:
            return result

        # HTML format: <p id="key">value</p>  (current MPC-BE default)
        for m in re.finditer(r'<p\s+id="([^"]+)">([^<]*)</p>', text):
            key, val = m.group(1), m.group(2).strip()
            # filepatharg is URL-encoded; also expose it as "filepath"
            if key == "filepatharg":
                result["filepath"] = urllib.parse.unquote(val)
            result[key] = val
        return result


def _ms_to_str(ms: int) -> str:
    if not ms:
        return "0:00"
    s = ms // 1000
    h, rem = divmod(s, 3600)
    m, sec = divmod(rem, 60)
    if h:
        return f"{h}:{m:02d}:{sec:02d}"
    return f"{m}:{sec:02d}"

File: app/MediaDownloader/MediaDownloader.Server/Clients/MpcClient.cs
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Web;

namespace MediaDownloader.Server.Clients;

/// <summary>
/// MPC-BE player status, parsed from /variables.html.
/// </summary>
public class MpcStatus
{
    private readonly Dictionary<string, string> _d;
    public bool Reachable { get; }

    public MpcStatus(Dictionary<string, string> data, bool reachable = true)
    {
        _d = data;
        Reachable = reachable;
    }

    public string File => _d.GetValueOrDefault("file", "") ?? _d.GetValueOrDefault("filepath", "");
    public string Filename => _d.GetValueOrDefault("filename", "") ??
        (File.Contains('\\') ? File.Split('\\')[^1] : File.Split('/')[^1]);
    public int State => int.TryParse(_d.GetValueOrDefault("state", "0"), out var v) ? v : 0;
    public bool IsPlaying => State == 2;
    public bool IsPaused => State == 1;
    public int PositionMs => int.TryParse(_d.GetValueOrDefault("position", "0"), out var v) ? v : 0;
    public int DurationMs => int.TryParse(_d.GetValueOrDefault("duration", "0"), out var v) ? v : 0;
    public string PositionStr => _d.GetValueOrDefault("positionstring", MsToStr(PositionMs));
    public string DurationStr => _d.GetValueOrDefault("durationstring", MsToStr(DurationMs));
    public int Volume => int.TryParse(_d.GetValueOrDefault("volumelevel", "100"), out var v) ? v : 100;
    public bool Muted
    {
        get
        {
            var val = _d.GetValueOrDefault("muted", "false");
            return val is "1" or "true" or "True";
        }
    }

    public Dictionary<string, object> ToDict() => new()
    {
        ["reachable"] = Reachable,
        ["file"] = File,
        ["filename"] = Filename,
        ["state"] = State,
        ["is_playing"] = IsPlaying,
        ["is_paused"] = IsPaused,
        ["position_ms"] = PositionMs,
        ["duration_ms"] = DurationMs,
        ["position_str"] = PositionStr,
        ["duration_str"] = DurationStr,
        ["volume"] = Volume,
        ["muted"] = Muted,
    };

    private static string MsToStr(int ms)
    {
        if (ms == 0) return "0:00";
        var s = ms / 1000;
        var h = s / 3600;
        var rem = s % 3600;
        var m = rem / 60;
        var sec = rem % 60;
        return h > 0 ? $"{h}:{m:D2}:{sec:D2}" : $"{m}:{sec:D2}";
    }
}

/// <summary>
/// MPC-BE web interface client. Mirrors server/clients/mpc_client.py.
/// </summary>
public class MpcClient
{
    private readonly string _baseUrl;
    private readonly ILogger? _logger;

    public MpcClient(string baseUrl, ILogger? logger = null)
    {
        _baseUrl = baseUrl.TrimEnd('/');
        _logger = logger;
    }

    public async Task<MpcStatus> GetStatusAsync(CancellationToken ct = default)
    {
        try
        {
            using var client = new HttpClient { Timeout = TimeSpan.FromSeconds(3) };
            var response = await client.GetStringAsync($"{_baseUrl}/variables.html", ct);
            var parsed = ParseVariables(response);
            return new MpcStatus(parsed, reachable: true);
        }
        catch
        {
            return new MpcStatus(new(), reachable: false);
        }
    }

    public async Task<bool> CommandAsync(int wmCommand, Dictionary<string, string>? extra = null, CancellationToken ct = default)
    {
        try
        {
            using var client = new HttpClient { Timeout = TimeSpan.FromSeconds(3) };
            var url = $"{_baseUrl}/command.html?wm_command={wmCommand}";
            if (extra != null)
            {
                foreach (var (k, v) in extra)
                    url += $"&{k}={Uri.EscapeDataString(v)}";
            }
            var response = await client.GetAsync(url, ct);
            return response.IsSuccessStatusCode;
        }
        catch { return false; }
    }

    public Task<bool> PlayPauseAsync(CancellationToken ct = default) => CommandAsync(887, ct: ct);
    public Task<bool> PlayAsync(CancellationToken ct = default) => CommandAsync(891, ct: ct);
    public Task<bool> PauseAsync(CancellationToken ct = default) => CommandAsync(892, ct: ct);
    public Task<bool> StopAsync(CancellationToken ct = default) => CommandAsync(888, ct: ct);
    public Task<bool> MuteAsync(CancellationToken ct = default) => CommandAsync(909, ct: ct);
    public Task<bool> VolumeUpAsync(CancellationToken ct = default) => CommandAsync(907, ct: ct);
    public Task<bool> VolumeDownAsync(CancellationToken ct = default) => CommandAsync(908, ct: ct);
    public Task<bool> SeekAsync(int positionMs, CancellationToken ct = default) =>
        CommandAsync(889, new() { ["position"] = positionMs.ToString() }, ct);

    public async Task<bool> PingAsync(CancellationToken ct = default)
    {
        try
        {
            using var client = new HttpClient { Timeout = TimeSpan.FromSeconds(2) };
            var response = await client.GetAsync($"{_baseUrl}/variables.html", ct);
            return response.IsSuccessStatusCode;
        }
        catch { return false; }
    }

    public async Task<bool> OpenFileAsync(string path, CancellationToken ct = default)
    {
        try
        {
            using var client = new HttpClient { Timeout = TimeSpan.FromSeconds(5) };
            var url = $"{_baseUrl}/command.html?wm_command=-1&path={Uri.EscapeDataString(path)}";
            var response = await client.GetAsync(url, ct);
            return response.IsSuccessStatusCode;
        }
        catch { return false; }
    }

    internal static Dictionary<string, string> ParseVariables(string text)
    {
        var stripped = text.Trim();

        // Try JSON first (newer MPC-BE versions)
        if (stripped.StartsWith('{'))
        {
            try
            {
                var json = JsonSerializer.Deserialize<Dictionary<string, JsonElement>>(stripped);
                if (json != null)
                    return json.ToDictionary(kv => kv.Key, kv => kv.Value.ToString());
            }
            catch { /* fall through */ }
        }

        var result = new Dictionary<string, string>();

        // Legacy JS format: OnVariable("key","value");
        foreach (Match m in Regex.Matches(text, @"OnVariable\(""([^""]+)"",""([^""]*)""\)"))
            result[m.Groups[1].Value] = m.Groups[2].Value;
        if (result.Count > 0) return result;

        // HTML format: <p id="key">value</p>
        foreach (Match m in Regex.Matches(text, @"<p\s+id=""([^""]+)"">([^<]*)</p>"))
        {
            var key = m.Groups[1].Value;
            var val = m.Groups[2].Value.Trim();
            if (key == "filepatharg")
                result["filepath"] = HttpUtility.UrlDecode(val);
            result[key] = val;
        }

        return result;
    }
}
